projects/chaos/src/state_handler.c

```diff
@@ -0,0 +1,53 @@
+#include "state_handler.h"
+
+#include <stddef.h>
+
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "log.h"
+#include "sequencer_fsm.h"
+
+#define STATE_HANDLER_EMPTY_DATA 0
+
+// CANRxHandlerCb
+static StatusCode prv_handle_power_state_msg(const CANMessage *msg, void *context,
+                                             CANAckStatus *ack_reply) {
+  (void)context;
+  EEPowerState power_state = NUM_EE_POWER_STATES;
+  CAN_UNPACK_POWER_STATE(msg, (uint8_t *)&power_state);
+
+  // Handle the power state.
+  switch (power_state) {
+    case EE_POWER_STATE_IDLE:
+      if (!status_ok(event_raise_priority(EVENT_PRIORITY_LOW, CHAOS_EVENT_SEQUENCE_IDLE, 0))) {
```

Comment:
Shouldn't the sequence transitions be high priority?

Reply:
So I believe this will not necessarily wait for a sequence to complete before changing to a new one. There should be a test for this if there isn't already.

I'm imagining a scenario where you're waiting for the relay to respond. In that timespan, if there are no other events in the normal priority FIFO, the sequence transition will be processed.

Reply:
Noted. I actually have an idea for slightly modifying the relay code to allow for cancelling requests which would solve this. I also need to modify the relay code anyways as there is a bug wherein the emergency state can get stuck in an infinite loop and never complete as the relay will just keep retrying. In the case of emergency I'm just going to assume the relays complete the first time and then keep a FIFO of any the failed and retry them forever until the fault is cleared or they succeed.

Reply:
Yeah, that seems much safer.